core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -341,9 +341,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
     if(targetSize > originalSize || targetSize < 0)
       throw new KafkaException("Attempt to truncate log segment to " + targetSize + " bytes failed, " +
                                " size of this log segment is " + originalSize + " bytes.")
-    channel.truncate(targetSize)
-    channel.position(targetSize)
-    _size.set(targetSize)
+    if (targetSize != channel.size()) {
+      channel.truncate(targetSize)
+      channel.position(targetSize)
```

Comment:
In case of the `FileMessageSet.close` it should be ok.
The other place where `truncateTo` is called is in `LogSegment.truncateTo()` - maybe @jkreps has some input to this.

If this remains unsure I could move the check to the `trim()` function in FileMessageSet.